docs/howtos/converting-a-component-to-uniffi.md

```diff
@@ -0,0 +1,287 @@
+to require some iteration to get it all working! Run `cargo check` to check for any
+compilation errors without having to do a full build.
+
+### Removing Protobuf Messages
+
+Passing rich structured data over the FFI is the most complex part of our hand-written bindings,
+and was previously done by [serializing data via Protocol Buffers](
+https://hacks.mozilla.org/2019/04/crossing-the-rust-ffi-frontier-with-protocol-buffers/).
+This is something that UniFFI tries to make as simple as possible.
+
+Start by locating the `<component_name>_msg_types.proto` file for the component. This file defines
+the structured messages that can be passed over the FFI, and you should see that they correspond
+to various types of structured data that the component wants to receive from, or return to,
+the foreign-language code.
+
+Find the places in your `.udl` interface the correspond to these message types and make sure
+they you've got a similarly-shaped `dictionary` or `enum` for each one. You should find that
+representing this structured data in UDL is simpler than protobuf in many cases - for example
+many of our `.protobuf` files need to use a separate `ExampleStructs` message in order to
+pass a list of `ExampleStruct` messages over the FFI, but in UniFFI this is represented
+directly as `sequence<ExampleStruct>`.
+
+Find the places in the Rust code that are using these message types to return structured data.
+In simple cases, you may be able to directly replace uses of `msg_types::ExampleStruct` with
+the corresponding `crate::ExampleStruct` from your public API.
+For more complex cases, you may find it helpful to define an `Into` mapping between the
+UniFFI dictionary/enum in the crate's public interface, and a more complex struct designed
+for internal use.
+
+Once you have replaced all uses of the `msg_types` structs in the Rust code:
+
+* Delete `./src/<component_name>_msg_types.proto`.
+* Delete `./src/mozilla.appservices.<component_name>.protobuf.rs`, which is generated from the `.proto` file.
+* Remote `prost` and `prost-derive` from the crate's dependencies.
+* Delete the crate from the list in `/tools/protobuf_files.toml`.
+
+If it happen to find that you've deleted the last crate from the list in `protobuf_files.toml`,
+congratulations! You've successfully removed protocol buffers from this repo entirely, and should
+file a bug to track the complete removal of protobuf from our tooling and dependency chain.
+
+## Document the Public API in the Rust code
+
+Write consumer-facing documentation on the public API in `lib.rs` using Rust's standard
+[rustdoc](https://doc.rust-lang.org/rustdoc/how-to-write-documentation.html) conventions
+and tools. The `fxa-client` crate may serve as a good example.
+
+You can view the generated documentation by running:
+
+```
+cargo doc --no-deps --open
+```
+
+In future, we intend to automatically extract documentation from the Rust code
+and make it easily available to consumers of the generated bindings.
```

Comment:
What I would like to write here is "run this cool script to copy the doc comments over into your UDL file" but I'm not sure that can be run by anyone but me currently. I'll take a look at what would be required to let that be runnable by someone else...

Reply:
@mhammond what do you think, is it worth me trying to make the automagical "copy docstrings from the Rust code into the UDL" functionality from https://github.com/mozilla/uniffi-rs/pull/416 into something more generally usable, or is that just getting ahead of ourselves?

Reply:
In one day? :) I think would be valuable, but not at all urgent. Maybe leaving a breadcrumb here would be worthwhile?